diff --git a/xilinx/pack.cc b/xilinx/pack.cc
index 0db5e2b..fe5f4de 100644
--- a/xilinx/pack.cc
+++ b/xilinx/pack.cc
@@ -469,28 +469,33 @@ void XilinxPacker::pack_constants()
     std::vector<IdString> dead_nets;
 
     std::vector<std::tuple<CellInfo *, IdString, bool>> const_ports;
+    std::unordered_set<IdString> dead_driver_cells;
 
     for (auto cell : sorted(ctx->cells)) {
         CellInfo *ci = cell.second;
         if (!tied_pins.count(ci->type))
             continue;
         auto &tp = tied_pins.at(ci->type);
@@ -501,15 +506,20 @@ void XilinxPacker::pack_constants()
     for (auto net : sorted(ctx->nets)) {
         NetInfo *ni = net.second;
-        if (ni->driver.cell != nullptr && ni->driver.cell->type == ctx->id("GND")) {
-            IdString drv_cell = ni->driver.cell->name;
-            for (auto &usr : ni->users) {
-                const_ports.emplace_back(usr.cell, usr.port, false);
-                usr.cell->ports.at(usr.port).net = nullptr;
-            }
-            dead_nets.push_back(net.first);
-            ctx->cells.erase(drv_cell);
-        } else if (ni->driver.cell != nullptr && ni->driver.cell->type == ctx->id("VCC")) {
-            IdString drv_cell = ni->driver.cell->name;
-            for (auto &usr : ni->users) {
-                const_ports.emplace_back(usr.cell, usr.port, true);
-                usr.cell->ports.at(usr.port).net = nullptr;
-            }
-            dead_nets.push_back(net.first);
-            ctx->cells.erase(drv_cell);
+        if (ni->driver.cell != nullptr &&
+            (ni->driver.cell->type == ctx->id("GND") || ni->driver.cell->type == ctx->id("VCC"))) {
+            bool cval = ni->driver.cell->type == ctx->id("VCC");
+            for (auto &usr : ni->users) {
+                const_ports.emplace_back(usr.cell, usr.port, cval);
+                usr.cell->ports.at(usr.port).net = nullptr;
+            }
+            dead_nets.push_back(net.first);
+            dead_driver_cells.insert(ni->driver.cell->name);
         }
     }
+
+    for (auto &drv_cell : dead_driver_cells)
+        ctx->cells.erase(drv_cell);
+    for (auto &dead_net : dead_nets)
+        ctx->nets.erase(dead_net);
 
     for (auto port : const_ports) {
